### PR TITLE
1. Unix domain sockets support. 2. Support for local and global working directories. 3. New APIs.

### PR DESCRIPTION
**Pull request by filipgolem**
1) Golem Unlimited Provider listens on a unix domain socket now.
https://github.com/golemfactory/gu-provider-ui-linux and
https://github.com/golemfactory/gu-provider-ui-mac connect to this socket.
2) Options that change working directories:
https://github.com/golemfactory/golem-unlimited/wiki/Working-Directories
3) New APIs for Golem Unlimited Provider:
https://github.com/golemfactory/golem-unlimited/wiki/APIs-Used-By-GU-Provider-UI

### CODE REVIEW

File: ethkey/src/lib.rs [hunk 1]
```diff
@@ -143,6 +143,11 @@ where
         crypto: secret.to_crypto(&password.into(), KEY_ITERATIONS)?,
         address: Some(Bytes(secret.public().address().to_vec())),
     };
+    let parent_dir = file_path.as_ref().parent().unwrap();
```
Comment (quilir) [inline ethkey/src/lib.rs]: What if a parent() call will return None?

File: gu-lan/src/module.rs [hunk 2]
```diff
@@ -119,4 +121,48 @@ impl Module for LanModule {
...
+        .map_err(|e| error!("error! {}", e))
+        .and_then(|r| {
+            Ok(HttpResponse::Ok().json(
+                r.unwrap_or(HashSet::new())
+                    .iter()
+                    .map(|instance| Reply {
+                        serv_type: instance.service(),
+                        host_name: instance.host.clone(),
+                        addr: format_addresses(&instance.addrs_v4, &instance.ports),
+                        desc: instance.txt.join(""),
```
Comment (quilir) [inline gu-lan/src/module.rs]: Maybe there is a better join string. Newline character is a more intuitive choice

File: gu-provider/src/permission.rs [hunk 3]
```diff
@@ -273,120 +373,120 @@ impl Module for PermissionModule {
...
+                        .into_future()
+                        .map_err(|_| actix_web::error::ErrorInternalServerError("bad format"))
+                        .and_then(move |node_or_auto| {
+                            set_node_status_future(
+                                turn_on,
+                                node_or_auto,
+                                hub.address,
+                                hub.host_name.clone(),
+                            )
+                            .map_err(|_| actix_web::error::ErrorInternalServerError(""))
```
Comment (quilir) [inline gu-provider/src/permission.rs]: Missing error message

File: gu-provider/src/permission.rs [hunk 4]
```diff
@@ -273,120 +373,120 @@ impl Module for PermissionModule {
...
+            });
+        })
+        .resource("{node_id}", |r| {
+            r.get().with_async(|path: Path<String>| {
+                extract_node_or_auto(path)
+                    .into_future()
+                    .map_err(|_| actix_web::error::ErrorInternalServerError("bad format"))
+                    .and_then(|node_or_auto| {
+                        get_node_status_future(node_or_auto)
+                            .map_err(|_| actix_web::error::ErrorNotFound(""))
```
Comment (quilir) [inline gu-provider/src/permission.rs]: Missing error message

File: gu-provider/src/permission.rs [hunk 5]
```diff
@@ -273,120 +373,120 @@ impl Module for PermissionModule {
...
+}
+
+fn config_methods<S: 'static>(scope: Scope<S>) -> Scope<S> {
+    scope
+        .resource("", |r| {
+            r.get().with_async(|q: Query<HashMap<String, String>>| {
+                (if q.contains_key("saved") {
+                    Ok(())
+                } else {
+                    Err(actix_web::error::ErrorBadRequest("Add ?saved to url"))
```
Comment (quilir) [inline gu-provider/src/permission.rs]:
```suggestion
                    Err(actix_web::error::ErrorBadRequest("Add \"?saved\" to url"))
```

File: gu-provider/src/server.rs [hunk 6]
```diff
@@ -214,12 +225,40 @@ impl<D: Decorator + 'static> Handler<InitServer<D>> for ProviderServer {
...
+                                    Err(e)
+                                });
+                                info!("Binding again to {:?}.", uds_path);
+                                tokio_uds::UnixListener::bind(&uds_path)
+                            } else {
+                                Err(e)
+                            }
+                        })
+                        .map_err(|e| {
+                            error!("Cannot bind to: {:?}. Please run with --user to create and use a unix domain socket in the user home directory", uds_path);
```
Comment (quilir) [inline gu-provider/src/server.rs]: 159 characters long line

**Review by quilir — APPROVED** [review]

File: ethkey/src/lib.rs (hunk 1 above)
Comment (filipgolem) [reply ethkey/src/lib.rs]: Fixed

File: gu-lan/src/module.rs (hunk 2 above)
Comment (filipgolem) [reply gu-lan/src/module.rs]: Fixed

File: gu-provider/src/permission.rs (hunk 4 above)
Comment (filipgolem) [reply gu-provider/src/permission.rs]: Fixed

File: gu-provider/src/permission.rs (hunk 3 above)
Comment (filipgolem) [reply gu-provider/src/permission.rs]: Fixed

File: gu-provider/src/server.rs (hunk 6 above)
Comment (filipgolem) [reply gu-provider/src/server.rs]: Fixed

**Review by prekucki — APPROVED** [review]

**Pull request merged**